include/onnxruntime/core/session/onnxruntime_run_options_config_keys.h

```diff
@@ -42,3 +42,8 @@ static const char* const kOrtRunOptionsConfigQnnPerfModePostRun = "qnn.htp_perf_
 
 // Set RPC control latency for QNN HTP backend
 static const char* const kOrtRunOptionsConfigQnnRpcControlLatency = "qnn.rpc_control_latency";
+
+// Set graph annotation id for CUDA EP. Use with enable_cuda_graph=true.
+// The value should be an integer. If the value is not set, ORT session only captures one cuda graph.
```

Comment:
I think it is better to have a default value so that every cuda graph has annotation. That could make the design simple. Otherwise, one graph without annotation, the other have annotations, and that design looks odd to me.

Reply:
Agree with the general sentiment. I had a similar comment too - https://github.com/microsoft/onnxruntime/pull/19636#discussion_r1503250933. Currently, the "annotation-less" graph would be the session level graph (not tied to any Run). If we could treat the session level graph as implicitly having annotation id as 0, then it may make things simpler. May need some more thinking though.

Reply:
There's a corner case on top of my head: if the default id is 0 and user explicitly pass id=0 what behavior should we expect to see?
1. override the existing graph with id=0
2. replay the graph with id=0
3. keep the graph with id=0 and transform the new id to another integer with some internal mapping and capture a new graph
4. we don't expect user to initialize a run_options without setting id and then pass an id in one session

Reply:
If user explicitly pass id=0, it is fine. The handling is just like other ID numbers: If the graph of id 0 not exist, capture and store the graph; otherwise, just replay it.

Reply:
I suggest support over-writing existing graph if annotation id is re-used. Do you see any major pitfalls with that ? Maybe we can log the event as a warning ?

Reply:
In my view, the last part _"otherwise, just replay the graph with id=0"_ may contradict with user's intention to capture a graph with id=0 if user previously does session run without an explicit id.

Reply:
Another idea maybe to "reserve" an annotation id for the session level graph and document that it cannot be used by users as an annotation id. That simplifies the code design and helps with the above concern.

Reply:
I think that all cuda graphs are session level. In one run, user could use ID (default 0) to trigger the capture (if the ID not captured yet) or replay (if the ID is captured), or ID=-1 to skip.

Reply:
I think Ye's concern (which seems valid to me) is if user didn't provide id (we still need to support legacy behavior) and we implcitly treat it as annotation id as 0 and then the user explicitly passes in 0 - what would the expected behavior ? I suggest documenting valid annotation ids are -1 (skip capture/replay) and 1...INT_MAX. '0' is invalid (and there can be gating check to ensure uses doesn't pass this). Internally we use annotation id as 0 to cache the graph if the user doesn't provide any annotation id (legacy behavior). Is this what you are suggesting too ?

Reply:
I think that's correct.